[ACR] Fix mypy errors

YalinLi0312 opened the pull request:
Resolves: https://github.com/Azure/azure-sdk-for-python/issues/20640

YalinLi0312 commented:
/azp run python - tables - ci

YalinLi0312 commented:
/azp run python - containerregistry - tests

YalinLi0312 commented:
/azp run python - containerregistry - ci

YalinLi0312 commented:
/azp run python - containerregistry - ci

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:
```diff
@@ -23,7 +23,7 @@ class AnonymousACRExchangeClient(object):
 
     def __init__(self, endpoint, **kwargs):  # pylint: disable=missing-client-constructor-parameter-credential
-        # type: (str, Dict[str, Any]) -> None
+        # type: (str, Any) -> None
```
Can we use Python3 type hints?

YalinLi0312 replied:
To make this PR easier to review, I didn't include the hints change. I'll create a separate for it.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:
```diff
@@ -36,25 +36,25 @@ def __init__(self, endpoint, **kwargs):  # pylint: disable=missing-client-constr
             service=service, scope=scope, refresh_token=refresh_token, grant_type=grant_type, **kwargs
         )
-        return access_token.access_token
+        return access_token.access_token # type: ignore
```
You shouldn't need to ignore this - is `self._client.authentication.exchange_acr_refresh_token_for_acr_access_token` type hinted correctly?

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:
```diff
@@ -36,25 +36,25 @@ def __init__(self, endpoint, **kwargs):  # pylint: disable=missing-client-constr
 
     def exchange_refresh_token_for_access_token(
-        self, refresh_token=None, service=None, scope=None, grant_type=TokenGrantType.PASSWORD, **kwargs
+        self, refresh_token, service, scope, grant_type, **kwargs
```
Is this a public method? Or only internal? Why were all the defaults removed?
Is there any possibility that one of these wouldn't be provided?

YalinLi0312 replied:
- It's internal used for now. 
- Will get `Bad request` when service or scope is None.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py:
```diff
@@ -22,11 +22,13 @@ class ContainerRegistryChallengePolicy(HTTPPolicy):
         self._credential = credential
         if self._credential is None:
-            self._exchange_client = AnonymousACRExchangeClient(endpoint, **kwargs)
+            # pylint: disable=line-too-long
```
You should just be able to add this to the one line that's too long.

YalinLi0312 replied:
Updated

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py:
```diff
@@ -22,11 +22,13 @@ class ContainerRegistryChallengePolicy(HTTPPolicy):
 
     def __init__(self, credential, endpoint, **kwargs):
-        # type: (TokenCredential, str, **Any) -> None
+        # type: (Optional[TokenCredential], str, **Any) -> None
```
Is this constructor public?

YalinLi0312 replied:
It is public. The hint is not correct, the credential could be None, we'll create `AnonymousACRExchangeClient` in this case.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:
```diff
@@ -143,7 +143,7 @@ def list_repository_names(self, **kwargs):
         n = kwargs.pop("results_per_page", None)
         last = kwargs.pop("last", None)
-        cls = kwargs.pop("cls", None)  # type: ClsType["_models.Repositories"]
+        cls = kwargs.pop("cls", None)  # type: str
```
`cls` isn't a string is it?

YalinLi0312 replied:
Have removed the type hint.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:
```diff
@@ -600,7 +600,7 @@ def update_manifest_properties(self, *args, **kwargs):
         :keyword bool can_write: Write permissions for a manifest.
         :rtype: ~azure.containerregistry.ArtifactManifestProperties
-        :raises: ~azure.core.exceptions.ResourceNotFoundError
+        :raises: ~azure.core.exceptions.ResourceNotFoundError, or TypeError when any arg is in invlid type.
```
You don't need to document a TypeError - this is already expected Python behaviour.

YalinLi0312 replied:
Updated

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:
```diff
@@ -621,10 +621,16 @@ def update_manifest_properties(self, *args, **kwargs):
         """
         repository = args[0]
+        if not isinstance(repository, str):
+            raise TypeError("Parameter repository should be type of string.")
```
Are these checks necessary?

YalinLi0312 replied:
Have removed the type checks.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:
```diff
@@ -621,10 +621,16 @@ def update_manifest_properties(self, *args, **kwargs):
         properties = None
         if len(args) == 3:
             properties = args[2]
+            if not isinstance(properties, ArtifactManifestProperties):
```
We shouldn't check for this type.

YalinLi0312 replied:
Have removed.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:
```diff
@@ -819,21 +835,29 @@ def upload_blob(self, repository, data, **kwargs):
         """
         try:
-            _, _, start_upload_response_headers = self._client.container_registry_blob.start_upload(
+            _, _, start_upload_response_headers = self._client.container_registry_blob.start_upload( # type: ignore
```
What were the errors here before you had to ignore them?

annatisch replied:
There are usually better ways to fix this than simply ignoring everything.

YalinLi0312 replied:
Have changed to just return headers.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py:
```diff
@@ -143,15 +144,6 @@ def fully_qualified_reference(self):
             _strip_alg(self._digest)
         )
 
-    def _to_generated(self):
```
We never send this object?

YalinLi0312 replied:
This is a duplicate one.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:
```diff
@@ -36,25 +36,25 @@ def __init__(self, endpoint, **kwargs):  # pylint: disable=missing-client-constr
         parsed_challenge["grant_type"] = TokenGrantType.PASSWORD
         return self.exchange_refresh_token_for_access_token(
-            None,
+            "",
```
This seems suspicious - what was the reason?

YalinLi0312 replied:
The value will pass to self._client.authentication.exchange_acr_refresh_token_for_acr_access_token() which is required for this parameter. I checked other languages, like in .Net is using empty string.

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py:
```diff
@@ -612,7 +615,7 @@ def update_manifest_properties(
         ...
 
-    @distributed_trace_async
+    @distributed_trace_async # type: ignore
```
This shouldn't be necessary?

YalinLi0312 replied:
Have removed

annatisch commented on sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_hello_world_async.py:
```diff
@@ -48,7 +48,7 @@ async def basic_sample(self):
                         print(tag.digest)
 
                     # [START delete_repository]
-                    await client.delete_repository(repository_name, "hello-world")
```
Was this sample previously broken?

YalinLi0312 replied:
delete_repository() only accept one positional value but we provide two here, it should broken.
But, YES- all samples, not only this one, are broken as lack of resources, they all need some images uploaded in advance. I have updated this step in the samples PR.

YalinLi0312 commented:
Updates: have addressed most of the comments except: 
1. pass `None` or `empty string` as in .Net for refresh_token while getting access token in anonymous access? 
2. should refresh_token be optional in rest API? 

We need to investigate them and will confirm with service team. cc: @annelo-msft

annelo-msft commented:
> We need to investigate them and will confirm with service team.

Yes, I plan to confirm what we're seeing today, and then let's discuss with the feature crew on our Teams thread prior to reaching out to the service team.  Will post what I learn here.

annelo-msft commented:
@YalinLi0312, here is [the line of code](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryChallengeAuthenticationPolicy.cs#L130) where .NET passes an empty string when obtaining a token for anonymous access.

There is not a place where we would pass a null value to this method, because we either pass `string.Empty` as shown here, or we pass the acrRefreshToken obtained from a service call or the token cache as shown [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryChallengeAuthenticationPolicy.cs#L115).

I found the guidance we received from the service team for obtaining anonymous access, and they shared:

> You can either use GET /oauth2/token without any auth credentials or POST /oauth2/token with `grant_type` as `password` with empty value to get a token for anonymous access.

Given this, I would recommend passing an empty value here rather than `None`, to comply with their recommendation.  If you do this, will it bypass the error you are seeing from your tools?

YalinLi0312 commented:
> @YalinLi0312, here is [the line of code](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryChallengeAuthenticationPolicy.cs#L130) where .NET passes an empty string when obtaining a token for anonymous access.
> 
> There is not a place where we would pass a null value to this method, because we either pass `string.Empty` as shown here, or we pass the acrRefreshToken obtained from a service call or the token cache as shown [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryChallengeAuthenticationPolicy.cs#L115).
> 
> I found the guidance we received from the service team for obtaining anonymous access, and they shared:
> 
> > You can either use GET /oauth2/token without any auth credentials or POST /oauth2/token with `grant_type` as `password` with empty value to get a token for anonymous access.
> 
> Given this, I would recommend passing an empty value here rather than `None`, to comply with their recommendation. If you do this, will it bypass the error you are seeing from your tools?

Thanks @annelo-msft! It no longer has errors in anonymous exchange client, but the error is still there in exchange client - `refresh_token` got by calling [this ](https://github.com/Azure/azure-sdk-for-python/blob/0e3da3968eda15796cb27f80b732e50c5826095d/sdk/containerregistry/azure-containerregistry/azure/containerregistry/_exchange_client.py#L77)function could be _None_,  while the expected `refresh_token` passing to [exchange_acr_refresh_token_for_acr_access_token()](https://github.com/Azure/azure-sdk-for-python/blob/0e3da3968eda15796cb27f80b732e50c5826095d/sdk/containerregistry/azure-containerregistry/azure/containerregistry/_exchange_client.py#L89) is expected to be a string. Do you know how to react when refresh_token is None?

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py:
```diff
@@ -77,7 +77,8 @@ def on_challenge(self, request, response, challenge):
         access_token = self._exchange_client.get_acr_access_token(challenge)
-        request.http_request.headers["Authorization"] = "Bearer " + access_token
+        if access_token is not None:
+            request.http_request.headers["Authorization"] = "Bearer " + access_token
```
This is a good bug fix :)

annatisch commented on sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_anonymous_exchange_client.py:
```diff
@@ -36,25 +36,20 @@ def __init__(  # pylint: disable=missing-client-constructor-parameter-credential
-    async def get_acr_access_token(self, challenge: str, **kwargs: Dict[str, Any]) -> str:
+    async def get_acr_access_token(self, challenge: str, **kwargs: Any) -> Optional[str]:
         parsed_challenge = _parse_challenge(challenge)
         parsed_challenge["grant_type"] = TokenGrantType.PASSWORD
```
We no longer need this line?

YalinLi0312 replied:
This is a good catch, I'll remove this line, thanks!

annatisch commented on sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_delete_images_async.py:
```diff
@@ -41,7 +48,7 @@ def __init__(self):
         load_registry()
         # Instantiate an instance of ContainerRegistryClient
-        async with ContainerRegistryClient(self.endpoint, self.credential, audience=self.audience) as client:
+        async with ContainerRegistryClient(self.endpoint, self.credential, audience=self.audience) as client: # type: ignore[arg-type]
```
What were these `ignore` statements in the samples for?

YalinLi0312 replied:
I don't remember, I'll remove it and check the pipeline result.

YalinLi0312 replied:
Interesting, the error is no longer there after remove one. I'll remove the rest.
